src/dotty/tools/dotc/util/Util.scala

```diff
@@ -0,0 +1,34 @@
+package dotty.tools.dotc.util
+import reflect.ClassTag
+
+object Util {
+  
+  /** The index `i` in `candidates.indices` such that `candidates(i) <= x` and
+   *  `candidates(i)` is closest to `x`, determined by binary search, or -1
+   *  if `x < candidates(0)`.
+   *  @param  hint   If between 0 and `candidates.length` use this
+   *                 as the first search point, otherwise use 
+   *                 `candidates.length/2`.
+   *  @pre   candidates is sorted
+   */
+  def bestFit(candidates: Array[Int], x: Int, hint: Int = -1): Int = {
+    def recur(lo: Int, hi: Int, mid: Int): Int =
+      if (x < candidates(mid)) 
+        recur(lo, mid - 1, (lo + mid - 1) / 2)
+      else if (mid + 1 < candidates.length && x >= candidates(mid + 1)) 
+        recur(mid + 1, hi, (mid + 1 + hi) / 2)
```

Comment:
This is prone to overflow (http://googleresearch.blogspot.com.au/2006/06/extra-extra-read-all-about-it-nearly.html). 

Of course we're unlikely to deal with a source file long enough to hit the problem :)